### PR TITLE
Filehash for path

### PR DESCRIPTION
**Pull request by zelima**
* [x] I've added tests to cover the proposed changes 

Changes proposed in this pull request:

- will add filehash to the path the resource should be placed in, if `add-filehash-to-path` set to true 
  - data/my_data.csv --> data/hash/my_data.csv 
- output dp.json will have filehash included into path appropriatelly 
- dump.to_path avoids rewiting files if already exist and `add-filehash-to-path`
 provided

### CODE REVIEW

File: datapackage_pipelines/lib/dump/dumper_base.py [hunk 1]
```diff
@@ -194,23 +195,24 @@ def copy_non_tabular_resources(self, datapackage):
...
                     stream = requests.get(url, stream=True).raw
                     stream.read = functools.partial(stream.read, decode_content=True)
                     shutil.copyfileobj(stream, tmp)
-                    filesize = tmp.tell()
-                    tmp.close()
-                    url = tmp.name
-                    delete = True
                 else:
-                    filesize = os.stat(url).st_size
+                    shutil.copyfile(url, tmp.name)
```
Comment (akariv) [inline datapackage_pipelines/lib/dump/dumper_base.py]: Why copy a local file to a temporary location?

File: datapackage_pipelines/utilities/resources.py [hunk 2]
```diff
@@ -31,6 +34,22 @@ def get_path(descriptor):
     return None
 
 
+def insert_hash_in_path(descriptor, hash):
```
Comment (akariv) [inline datapackage_pipelines/utilities/resources.py]: This should be part of DumperBase, not in utilities.

File: datapackage_pipelines/utilities/resources.py [hunk 3]
```diff
@@ -31,6 +34,22 @@ def get_path(descriptor):
...
+    path = descriptor.get('path')
+    if isinstance(path, str):
+        path_list = path.split('/')
```
Comment (akariv) [inline datapackage_pipelines/utilities/resources.py]: no, this is not portable.
use `os.path.dirname()` and `os.path.basename()`

File: datapackage_pipelines/utilities/resources.py [hunk 4]
```diff
@@ -31,6 +34,22 @@ def get_path(descriptor):
...
+        hashed = os.path.join(hash, path_list.pop(-1))
+        path_list.append(hashed)
+        descriptor['path'] = '/'.join(path_list)
+    if isinstance(path, list):
+        if len(path) > 0:
+            path = path[0]
```
Comment (akariv) [inline datapackage_pipelines/utilities/resources.py]: There's some unnecessary code duplication here.
We can have path to be a string always (regardless of its initial form)

File: datapackage_pipelines/lib/dump/dumper_base.py [hunk 5]
```diff
@@ -263,3 +263,16 @@ def handle_resource(self, resource, spec, _, datapackage):
...
             return resource
+
+    @staticmethod
+    def hash_handler(tfile):
+        tfile.seek(0)
+        hasher = hashlib.md5()
+        data = 'x'
+        while len(data) > 0:
+            data = tfile.read(1024)
+            try:
```
Comment (akariv) [inline datapackage_pipelines/lib/dump/dumper_base.py]: better to test for type (str/bytes) than to rely on exception handling

File: README.md [hunk 6]
```diff
@@ -677,6 +677,7 @@ _Parameters_:
 - `format` - Specifies the type of output files to be generated (if `force-format` is true): `csv` (the default) or `json`
 - `handle-non-tabular` - Specifies whether non tabular resources (i.e. resources without a `schema`) should be dumped as well to the resulting datapackage.
     (See note below for more details)
+- `add-filehash-to-path`: Specifies whether to include file hash into the resource path. Defaults to `False`
```
Comment (akariv) [inline README.md]: Please elaborate a bit (i.e. which hash, how it's added to the path).

File: datapackage_pipelines/lib/dump/dumper_base.py (hunk 1 above)
Comment (zelima) [reply datapackage_pipelines/lib/dump/dumper_base.py]: @akariv thought would be better than loading it into memory (as we have to open to get hash)

Comment (zelima) [reply datapackage_pipelines/lib/dump/dumper_base.py]: Hm... realised that anyway have to do that :)

File: datapackage_pipelines/lib/dump/dumper_base.py [hunk 7]
```diff
@@ -271,8 +289,8 @@ def hash_handler(tfile):
         data = 'x'
         while len(data) > 0:
             data = tfile.read(1024)
-            try:
+            if isinstance(data, str):
                 hasher.update(data.encode('utf8'))
-            except AttributeError:
+            if isinstance(data, bytes):
```
Comment (akariv) [inline datapackage_pipelines/lib/dump/dumper_base.py]: elif

**Pull request merged**